### ISSUE TITLE
cmd/cover: inconsistent treatment of comments

### ISSUE THREAD

**Issue by alandonovan** [golang/go#22545]
Run coverage on the following program using these commands:

```
$ cd $GOROOT/src/test
$ cat test.go
package test

func f() {
        if true {
                println("A")
        }
        // comment
        if true {
                // comment
                if false {
                        println("B")
                }
        } else {
                // comment
                println("D")
        }
}
$ cat test_test.go 
package test

import "testing"

func Test(t *testing.T) { f() }
$ go test -coverprofile=c.out test
ok      test    0.015s  coverage: 66.7% of statements
$ command go tool cover -html=c.out
```

This is the rendered HTML output:

![coverage](https://user-images.githubusercontent.com/8433140/32339357-fac47eaa-bfcd-11e7-9d90-c07be51c153c.png)

Observe that the first comment is grey, not green, even though it was covered. Is this a simple bookkeeping error, or is there a design reason why we shouldn't consider the entire span from `func f() {` up to `if false` as covered, and render it green?

(This is Google internal issue 68650370.)

**Comment by gregoryv** [comment]
> I've been observing this and similar. Comments and function headers (the bit before the opening brace) should be grey, because they are not executable statements. (I shouldn't decrease my code's coverage simply by adding a bunch more comments, for instance.)

@dsymonds adding more comments to the uncovered section in your example does not decrease test coverage, the numbers are the same even if there are more non-green lines.

![comments](https://user-images.githubusercontent.com/779941/71807346-64437d80-306b-11ea-835c-762faabef4f2.png)

The colours are however misleading as you say, some are green and others are not. I've been looking into this, not in the "cover" tool but in a [variant](https://github.com/gregoryv/uncover)(based on the original) which outputs the result to a terminal.

However I disagree that the function signature should be grey, it is called during execution isn't it?
Comments are also grey(sometimes) but in the view of a coverage analysis I think they are part of the path between two statements and execution does cover that path.

In my variation I actually changed this to only use green and red colours which resulted in

![uncover-colors](https://user-images.githubusercontent.com/779941/71806797-06fafc80-306a-11ea-8c82-3861b0b87fca.png)

Changes can be summarized as

- signature is green if it has > 0% coverage
- function open and close braces are green
- all comments crossed by an executed path within a function are green

Benefit of this is the reader only has to scan for the red colour, grey and green basically mean the same within a function context.
To summarize; I would like to simplify colouring to only display red and green within a function context.

**Comment by cb80** [comment]
There is more to it. Such cases lead to issues with tools like SONARQUBE reporting on lines - not on the number of covered statements.

Example code:
```go
package main

import "fmt"

func greeting(name string) string {
	if name != "" {
		return "Hello " + name
	}

	// define message
	msg := "Forgot your name?"

	// return
	return msg
}

func main() {
	fmt.Println(greeting("Jim"))
}
```

And the following test:
```go
func TestGreet(t *testing.T) {
	got := greeting("Jim")
	if got != "Hello Jim" {
		t.Errorf("Expected Hello Jim, but got %s", got)
	}
}
```

This produces the following cover profile:
```
mode: set
bar/main.go:5.35,6.16 1 1
bar/main.go:6.16,8.3 1 1
bar/main.go:11.2,14.12 2 0
bar/main.go:17.13,19.2 1 0
```

SONARQUBE interprets line `bar/main.go:11.2,14.12 2 0` that those code lines are not covered (including the comment):
```go
	msg := "Forgot your name?"

	// return
	return msg
```

It would be great if Go can write a new line to the cover profile whenever there is a comment. In that case this would be:
```
mode: set
bar/main.go:5.35,6.16 1 1
bar/main.go:6.16,8.3 1 1
bar/main.go:11.2,11.28 1 0
bar/main.go:14.2,14.12 1 0
bar/main.go:17.13,19.2 1 0
```

**Comment by rudy-regazzoni-sonarsource** [comment]
We observed also such issue being reported by SonarQube users.

The current behavior consider some comments as being part of covered/uncovered code, which doesn't make much sense as comments are not executable statements.
While we could consider at first that this is something trivial, it actually lead subsequent tooling to work with improper data.

The produced cover profile is used in SonarQube to compute the percentage of code coverage, which ultimately will make the Quality Gate green or not.
You may see it coming: we got some case of the Quality Gate preventing the developers from merging their PR, because they had too much of uncovered comments.
In worst case, the dev can be tempted to reduce or even remove their comment to make their Quality Gate green and merge their PR, which goes against the logic of ensuring good code quality.

As mentioned above, I would strongly suggest to modify the go test coverage report to exclude comment only lines from it.

**Comment by seankhliao** [comment]
Given how long this issue has been open, if this is important to SonarSource, then you're welcome to submit a change to resolve the issue.
